### PR TITLE
Webrpc SSE streaming examples

### PR DESCRIPTION
**Pull request by VojtechVitek**
Heavily based on great work at https://github.com/webrpc/webrpc/tree/next-streaming branch by @pkieltyka. It follows the same Go<->Go API and differs in the following implementation details:

- Uses `application/x-ndjson` instead of `application/stream+json`
    - The Newline-Delimited JSON format ([NDJSON](https://ndjson.org/)) has been adopted by some major projects:
        - [ElasticSearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html), Kibana, Grafana, Elastic Beats
        - [Fluentd](https://docs.fluentd.org/output/http)
        - [WireShark](https://github.com/wireshark/wireshark/blob/master/doc/tshark.adoc)
        - [Spring framework](https://docs.spring.io/spring-framework/reference/web/webflux/reactive-spring.html#webflux-codecs-streaming) - [PR](https://github.com/spring-projects/spring-framework/issues/21283#issuecomment-515038827)
        - [ClickHouse](https://github.com/ClickHouse/ClickHouse/pull/32223)
        - etc.
    - The keep-alive PING message is simply a newline `\n`
    - The DATA message is not wrapped into another JSON object (`{"data": { ... }`), thus eliminating extra marshaling/unmarshaling costs in the hot path and making the response more readable and ingestible by other NDJSON-compatible services
    - The ERROR message is wrapped in a `{"webrpcError": { ... }}` object, making it unique-enough & easy to decode by clients
    - Go server uses `json.Encoder()` and client uses blocking `json.Decoder()` as the only I/O operation
- Uses native Go stdlib and browser support for `Transfer-Encoding: chunked` header
   - Go `http.Server` automatically switches to chunked encoding when `Content-Length` header is missing
   - The chunk-length integer doesn't show up in the response payload anymore, as it's "hidden" by HTTP/1.1
   - No need to use `ChunkDecoder` in TypeScript client, as it can simply read new lines from the response body and call `JSON.parse()` on each line's text
- The TypeScript StreamReader API has been extended to support `AbortSignal` argument, same as in regular RPC endpoints https://github.com/webrpc/gen-typescript/pull/9

## RIDL
```diff
 service Chat
   - SendMessage(authorName: string, text: string)
+  - SubscribeMessages() => stream (messages: Message)
```

## Go server interface
```diff
 type Chat interface {
 	SendMessage(ctx context.Context, authorName string, text string) error
+	SubscribeMessages(ctx context.Context, stream SubscribeMessagesStreamWriter) error
 }
```

## Go client interface
```diff
 type ChatClient interface {
 	SendMessage(ctx context.Context, authorName string, text string) error
+	SubscribeMessages(ctx context.Context) (SubscribeMessagesStreamReader, error)
 }
```

## CURL

```diff
 $ curl -v -X POST -H "Content-Type: application/json" -H "Accept: application/x-ndjson" --data '{}' http://localhost:4848/rpc/Chat/SubscribeMessages
 *   Trying 127.0.0.1:4848...
 * Connected to localhost (127.0.0.1) port 4848 (#0)
 > POST /rpc/Chat/SubscribeMessages HTTP/1.1
 > Host: localhost:4848
 > User-Agent: curl/7.79.1
 > Content-Type: application/json
 > Accept: application/x-ndjson
 > Content-Length: 23
 > 
 * Mark bundle as not supporting multiuse
 < HTTP/1.1 200 OK
 < Cache-Control: no-store
 < Connection: keep-alive
+< Content-Type: application/x-ndjson
 < X-Content-Type-Options: nosniff
 < Date: Tue, 28 Nov 2023 09:16:17 GMT
+< Transfer-Encoding: chunked
 < 
 {"message":{"id":17207198650781032141,"text":"Message 0","authorName":"Alice","createdAt":"0001-01-01T00:00:00Z"}}
 {"message":{"id":8335195124288779737,"text":"Message 1","authorName":"Alice","createdAt":"0001-01-01T00:00:00Z"}}
 {"message":{"id":3353724253333525515,"text":"Message 2","authorName":"Alice","createdAt":"0001-01-01T00:00:00Z"}}
 {"message":{"id":9597642916308127485,"text":"Message 3","authorName":"Alice","createdAt":"0001-01-01T00:00:00Z"}}
 {"message":{"id":11519856362397588428,"text":"Message 4","authorName":"Alice","createdAt":"0001-01-01T00:00:00Z"}}
+{"webrpcError":{"error":"ConnectionTooLong","code":100,"msg":"Connection too long. Terminated.","cause":"timed out after 2s","status":400}}
* Connection #0 to host localhost left intact
```

## Golang server

```go
func (s *Server) SubscribeMessages(ctx context.Context, stream proto.SubscribeMessagesStreamWriter) error {
	ctx, _ = context.WithTimeout(ctx, 3*time.Second) // will trigger ErrConnectionTooLong error after 3s

	for i := 0; i < 10; i++ {
		select {
		case <-ctx.Done():
			switch err := ctx.Err(); err {
			case context.Canceled:
				return fmt.Errorf("client disconnected") // useful for server logs
			default:
				return proto.ErrConnectionTooLong.WithCause(fmt.Errorf("timed out after %vs", serverTimeoutSec))
			}

		case <-time.After(time.Duration(rand.Intn(1000)) * time.Millisecond):
			// Simulate async work. Delay each message by 0-1000ms.
		}

		msg := &proto.Message{
			Id:         rand.Uint64(),
			Text:       fmt.Sprintf("Message %v", i),
		}

		if err := stream.Write(msg); err != nil {
			return err
		}
	}

	return nil
}
```

## Golang client
```go
func subscribe(ctx context.Context) {
	stream, err := client.SubscribeMessages(ctx)
	if err != nil {
		return err
	}

	for {
		msg, err := stream.Read()
		if err != nil {
			if errors.Is(err, proto.ErrWebrpcStreamFinished) {
				break
			}
			return err
		}
		fmt.Println(msg.Text)
	}
	
	return nil
}
```

Automatically re-connect to a lost SSE stream (ie. due to a network error):
```go
ctx := context.Background() // try to keep SSE connection open forever

for {
	err := subscribe(ctx)
	if err != nil {
		if errors.Is(err, proto.ErrWebrpcStreamLost)
			continue // re-connect right away
		}
		
		// handler other errors sent either by server or net/http pkg
		// CAUTION: re-connect only with exponential backoff here
	}
	
	return nil // stream successfully finished
}
```

## TypeScript client

Note: We didn't choose to support browser SSE [EventStream](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events) API, as it's too limited. It can't use `POST` method, set custom headers or send request body. However, we got inspired by the EventStream API and also by Azure's [fetch-event-source polyfill](https://github.com/Azure/fetch-event-source).

```ts
const onMessage = (message: SubscribeMessagesReturn) => {
  console.log("onMessage()", message);
};

const onError = (error: WebrpcError, reconnect: () => Promise<void>) => {
  console.error("onError()", error);

  if (error instanceOf WebrpcStreamLostError) {
    reconnect()
  }
};

const onOpen = () => {
  console.log("onOpen()");
};

const onClose = () => {
  console.log("onClose()");
};

const controller = new AbortController();
const abortSignal = controller.signal;

// Subscribe to SSE stream
try {
  await api.subscribeMessages(
    { username },
    { onMessage, onError, onOpen, onClose, signal: abortSignal }
  );
} catch (error) {
  console.error(error)
}
```

### CODE REVIEW

File: _examples/golang-sse/chat_test.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+	ctx := context.Background()
+
+	stream, err := client.SubscribeMessages(ctx, 2)
+	require.Nil(t, err)
+
+	for {
+		msg, err := stream.Read()
+		if err != nil {
+			assert.ErrorIs(t, err, proto.ErrConnectionTooLong)
```
Comment (pkieltyka) [inline _examples/golang-sse/chat_test.go:106]: the name `ErrConnectionTooLong` is a bit strange.. how about.. `ErrConnectionTimedOut` ..? or.. `ErrConnectionTimeout`

File: _examples/golang-sse/webapp/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import { effect, signal } from "@preact/signals-core";
+import { Chat, Message, SubscribeMessagesReturn, WebrpcError } from "./rpc";
+import "./style.css";
+
+//Create client
```
Comment (pkieltyka) [inline _examples/golang-sse/webapp/src/main.ts:5]: minor style thing, please use space after the `//` .. so.. `// Create client`

File: _examples/golang-sse/webapp/src/rpc.ts [hunk 3]
```diff
@@ -0,0 +1,477 @@
+/* eslint-disable */
```
Comment (pkieltyka) [inline _examples/golang-sse/webapp/src/rpc.ts:1]: can you rename this file to `rpc.gen.ts` I prefer to use `.gen.xxx` for filenames whenever code is generated so its obvious

File: _examples/golang-sse/webapp/src/rpc.ts [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+export const WebRPCSchemaVersion = "v1.0.0";
+
+// Schema hash generated from your RIDL schema
+export const WebRPCSchemaHash = "a63a64964867e7edb412b0400e36220b6057c41c";
+
+//
+// Types
+//
+
+//Global types
```
Comment (pkieltyka) [inline _examples/golang-sse/webapp/src/rpc.ts:21]: add space after `//`

**Review by pkieltyka — COMMENTED** [review]
Nice work!

File: _examples/golang-sse/chat_test.go (hunk 1 above)
Comment (VojtechVitek) [reply _examples/golang-sse/chat_test.go:106]: This is an example of a custom schema error. I'll rename it to `ErrServerShuttingDown` or something. 👍

File: _examples/golang-sse/webapp/src/rpc.gen.ts [hunk 5]
```diff
@@ -511,7 +514,9 @@ export interface WebrpcOptions {
 
 export interface WebrpcStreamOptions<T> extends WebrpcOptions {
   onMessage: (message: T) => void;
-  onError: (error: WebrpcError, reconnect?: () => void) => void;
+  onError:
+    | ((error: WebrpcError) => void)
+    | ((error: WebrpcError, reconnect: () => void) => void);
```
Comment (VojtechVitek) [inline _examples/golang-sse/webapp/src/rpc.gen.ts:519]: @kolberg-dev Should we pass `reconnect()` at all times; and throw an error if someone tries to reconnect on abort signal (ie. `WebrpcRequestCanceled` error)?

We could document that `reconnect()` shouldn't be used when the stream is explicitly canceled.

Comment (kolberg-dev) [reply _examples/golang-sse/webapp/src/rpc.gen.ts:527]: https://github.com/webrpc/webrpc/pull/237/commits/89881be53915a71b2157ccc89736eb64265105b5
I dont think this should be a webrpcerror, we should break the app during dev

**Pull request merged**